metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -624,50 +617,71 @@ public String toString() {
+                            snapshotRegistry.deleteSnapshotsUpTo(
+                                Math.min(offset, snapshotGeneratorManager.snapshotEpoch()));
+
                         } else {
-                            log.debug("Replaying commits from the active node up to " +
-                                "offset {}.", offset);
+                            // If the controller is a standby, replay the records that were
+                            // created by the active controller.
+                            if (log.isDebugEnabled()) {
+                                if (log.isTraceEnabled()) {
+                                    log.trace("Replaying commits from the active node up to " +
+                                        "offset {}: {}.", offset, messages.stream()
+                                        .map(ApiMessageAndVersion::toString)
+                                        .collect(Collectors.joining(", ")));
+                                } else {
+                                    log.debug("Replaying commits from the active node up to " +
+                                        "offset {}.", offset);
+                                }
+                            }
+                            for (ApiMessageAndVersion messageAndVersion : messages) {
+                                replay(messageAndVersion.message(), -1, offset);
+                            }
                         }
+                        lastCommittedOffset = offset;
                     }
-                    for (ApiMessage message : messages) {
-                        replay(message, -1, offset);
-                    }
-                } else {
-                    // If the controller is active, the records were already replayed,
-                    // so we don't need to do it here.
-                    log.debug("Completing purgatory items up to offset {}.", offset);
-
-                    // Complete any events in the purgatory that were waiting for this offset.
-                    purgatory.completeUpTo(offset);
-
-                    // Delete all the in-memory snapshots that we no longer need.
-                    // If we are writing a new snapshot, then we need to keep that around;
-                    // otherwise, we should delete up to the current committed offset.
-                    snapshotRegistry.deleteSnapshotsUpTo(
-                        Math.min(offset, snapshotGeneratorManager.snapshotEpoch()));
+                } finally {
+                    reader.close();
                 }
-                lastCommittedOffset = offset;
             });
         }
 
         @Override
-        public void handleNewLeader(MetaLogLeader newLeader) {
-            if (newLeader.nodeId() == nodeId) {
-                final long newEpoch = newLeader.epoch();
+        public void handleSnapshot(SnapshotReader<ApiMessageAndVersion> reader) {
```

Comment:
I think there's some ambiguity in how this API is named. "Handling" a snapshot could mean either writing one or reading it. Judging from the parameter names here, I'm guessing that we want to read the snapshot here.

My understanding is that this API will be called if the standby controller falls behind too far, OR right after it starts up and before any other records are read. Does that match up with your understanding?

If that's true, then we should be able to put the snapshot loading stuff that currently exists here, with a TODO for implementing snapshot loading AFTER the controller has already been initialized (which will involve zeroing out all the existing data structures and in-memory snapshots...) I think we should do that since I don't want this patch to move us backwards in snapshot support.

Reply:
> If that's true, then we should be able to put the snapshot loading stuff that currently exists (in the constructor) here, with a TODO for implementing snapshot loading AFTER the controller has already been initialized (which will involve zeroing out all the existing data structures and in-memory snapshots...) I think we should do that since I don't want this patch to move us backwards in snapshot support.

We discussed this in another comment thread in this PR but this PR has a partial implementation of snapshot loading in the controller. I am going to implement snapshot re-loading as part of https://issues.apache.org/jira/browse/KAFKA-12787.